### PR TITLE
EVG-5638: rehome evergreen's s3 useage ontop of pail

### PR DESCRIPTION
**Pull request by julianedwards**

### CODE REVIEW

File: service/api_plugin_s3copy.go [hunk 1]
```diff
@@ -83,9 +83,25 @@ func (as *APIServer) s3copyPlugin(w http.ResponseWriter, r *http.Request) {
...
+		Region:      region,
+		Name:        s3CopyReq.S3SourceBucket,
+		Permission:  "public-read",
+	}
+	srcBucket, err := pail.NewS3MultiPartBucket(srcOpts)
+	if err != nil {
+		grip.Error(errors.Wrap(err, "S3 copy failed, could not establish connection to source bucket"))
+	}
+	destOpts := pail.S3Options{
+		Credentials: credentials.NewStaticCredentials(s3CopyReq.AwsKey, s3CopyReq.AwsSecret, ""),
```
Comment (bsamek) [inline service/api_plugin_s3copy.go]: Credentials

File: command/s3_get.go [hunk 2]
```diff
@@ -229,20 +220,28 @@ func (c *s3get) get(ctx context.Context) error {
...
 		return errors.Wrapf(err, "problem extracting %s from archive", c.RemoteFile)
 	}
 
 	return nil
 }
+
+func (c *s3get) createPailBucket() error {
+	opts := pail.S3Options{
+		Credentials: credentials.NewStaticCredentials(c.AwsKey, c.AwsSecret, ""),
+		Region:      "us-east-1",
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply command/s3_get.go]: Unfortunately not :(

File: command/s3_put.go [hunk 3]
```diff
@@ -382,3 +381,16 @@ func (s3pc *s3put) attachFiles(ctx context.Context, comm client.Communicator, lo
 	logger.Execution().Info("API attach files call succeeded")
 	return nil
 }
+
+func (c *s3put) createPailBucket() error {
+	opts := pail.S3Options{
+		Credentials: credentials.NewStaticCredentials(c.AwsKey, c.AwsSecret, ""),
+		Region:      "us-east-1",
+		Name:        c.Bucket,
+		Permission:  c.Permissions,
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply command/s3_put.go]: ack

File: service/api_plugin_s3copy.go [hunk 4]
```diff
@@ -7,21 +7,21 @@ import (
...
-	"github.com/goamz/goamz/s3"
+	"github.com/evergreen-ci/pail"
 	"github.com/mongodb/grip"
 	"github.com/pkg/errors"
 )
 
 const (
 	s3CopyRetrySleepTimeSec = 5
 	s3CopyRetryNumRetries   = 5
+	region                  = "us-east-1"
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply service/api_plugin_s3copy.go]: nope :(

File: service/api_plugin_s3copy.go [hunk 5]
```diff
@@ -83,9 +83,25 @@ func (as *APIServer) s3copyPlugin(w http.ResponseWriter, r *http.Request) {
...
 
 	// Now copy the file into the permanent location
-	auth := &aws.Auth{
-		AccessKey: s3CopyReq.AwsKey,
-		SecretKey: s3CopyReq.AwsSecret,
+	srcOpts := pail.S3Options{
+		Credentials: credentials.NewStaticCredentials(s3CopyReq.AwsKey, s3CopyReq.AwsSecret, ""),
+		Region:      region,
+		Name:        s3CopyReq.S3SourceBucket,
+		Permission:  "public-read",
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply service/api_plugin_s3copy.go]: nope :(

File: command/s3_get.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (bsamek) [reply command/s3_get.go]: Can you use either https://github.com/evergreen-ci/evergreen/blob/master/vendor/github.com/aws/aws-sdk-go/aws/endpoints/defaults.go#L29 or a constant you create in Evergreen code?

File: thirdparty/s3_test.go [hunk 6]
```diff
@@ -1,123 +0,0 @@
-package thirdparty
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply thirdparty/s3_test.go]: these tests are equivalent to the tests in pail... would you like additional tests here in evergreen?

File: command/s3_get.go [hunk 7]
```diff
@@ -229,20 +226,34 @@ func (c *s3get) get(ctx context.Context) error {
...
 	return nil
 }
+
+func (c *s3get) createPailBucket() error {
+	opts := pail.S3Options{
+		Credentials: credentials.NewCredentials(&credentials.StaticProvider{
+			Value: credentials.Value{
+				AccessKeyID:     c.AwsKey,
+				SecretAccessKey: c.AwsSecret,
+			}}),
```
Comment (tychoish) [inline command/s3_get.go]: Can we do this using the new pail structures so we don't need to reference the credentials structure?

File: command/s3_get.go [hunk 8]
```diff
@@ -195,30 +212,10 @@ func (c *s3get) getWithRetry(ctx context.Context, logger client.LoggerProducer)
...
 // Fetch the specified resource from s3.
 func (c *s3get) get(ctx context.Context) error {
-	// get the appropriate session and bucket
-	auth := &aws.Auth{
-		AccessKey: c.AwsKey,
-		SecretKey: c.AwsSecret,
-	}
-
-	client := util.GetHTTPClient()
-	defer util.PutHTTPClient(client)
```
Comment (tychoish) [inline command/s3_get.go]: how does pail get an http client? we should make sure that we don't leak connections by switching to pail.

File: service/api_plugin_s3copy.go [hunk 9]
```diff
@@ -82,9 +84,33 @@ func (as *APIServer) s3copyPlugin(w http.ResponseWriter, r *http.Request) {
 	}
 
 	// Now copy the file into the permanent location
-	auth := &aws.Auth{
-		AccessKey: s3CopyReq.AwsKey,
-		SecretKey: s3CopyReq.AwsSecret,
+	srcOpts := pail.S3Options{
+		Credentials: credentials.NewCredentials(&credentials.StaticProvider{
```
Comment (tychoish) [inline service/api_plugin_s3copy.go]: avoid calling out to the credentials package directly.

File: command/s3_get.go [hunk 10]
```diff
@@ -2,16 +2,16 @@ package command
...
 	"path/filepath"
 	"time"
 
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
 	"github.com/evergreen-ci/evergreen/model"
 	"github.com/evergreen-ci/evergreen/rest/client"
-	"github.com/evergreen-ci/evergreen/thirdparty"
 	"github.com/evergreen-ci/evergreen/util"
-	"github.com/goamz/goamz/aws"
```
Comment (tychoish) [inline command/s3_get.go]: can we also unvender goamz as part of this change?

Comment (julianedwards) [reply command/s3_get.go]: done

**Pull request merged**